doc/dev/templates.md

```diff
@@ -0,0 +1,143 @@
+depends on each adapter’s category. All adapters of a given category receive the same type of data. For example,
+metric adapters all receive metrics at runtime.
+
+Mixer determines the format of data that each adapter receive using *templates*. A template is a simple protobuf message
+which lays out the kind of data that a specific category of adapter will be given at runtime. The template determines
+the data adapters receive at runtime, as well as the instances operators must create in order to use an adapter.
+The [Adapter Developer's Guide](./adapters.md) explains how templates are automatically transformed into Go
+artifacts that can be used by adapter developers, and into config definitions that can be used by operators. 
+
+Mixer includes a number of [canonical templates](https://github.com/istio/mixer/tree/master/template) which cover
+most of the anticipated workloads that Mixer is expected to be used with. However, the set of supported templates can
+readily be extended in order to support emerging usage scenarios. Note that it’s preferable to use existing templates
+when possible as it tends to deliver a better end-to-end story for the ecosystem. 
+
+This document describes the simple rules used to create templates for Mixer.
+The [Adapter Developer’s Guide](https://github.com/istio/mixer/blob/master/doc/dev/adapters.md) describes how to use templates to
+create adapters.
+
+## Template format
+
+Templates are expressed using the protobuf syntax. They are generally fairly simple data structures. An an example, here is the `listentry` template:
+
+<pre>
+syntax = "proto3";
+
+package listEntry;
+
+import "mixer/v1/template/extensions.proto";
+
+option (istio.mixer.v1.template.template_variety) = TEMPLATE_VARIETY_CHECK;
+
+// ListEntry is used to verify the presence/absence of a string
+// within a list.
+//
+// When writing the configuration, the value for the fields associated with this template can either be a
+// literal or an [expression](https://istio.io/docs/reference/config/mixer/expression-language.html). Please note that if the datatype of a field is not istio.mixer.v1.config.descriptor.ValueType,
+// then the expression's [inferred type](https://istio.io/docs/reference/config/mixer/expression-language.html#type-checking) must match the datatype of the field.
+//
+// Example config:
+// 
+// apiVersion: "config.istio.io/v1alpha2"
+// kind: listentry
+// metadata:
+//   name: appversion
+//   namespace: istio-config-default
+// spec:
+//   value: source.labels["version"]
+message Template {
+    // Specifies the entry to verify in the list.
+    string value = 1;
+}
+</pre>
+
+The interesting parts of this definition are:
+
+- **Package Name**. The package name determines the name by which the template will be known, both by
+adapter authors and operators using adapters that implement the template. 
+The name should be written in camelCase.
```

Comment:
The package name is also used to generate Go interfaces and types that will be used by Adapters. The name should be written in camelCase in order to create more readable named Go artifacts